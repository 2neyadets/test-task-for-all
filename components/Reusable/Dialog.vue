<template lang="pug">
  transition(name="slide")
    .q-dialog.fullscreen.no-pointer-events(v-if="value")
      .q-dialog__backdrop.fixed-full(@click="$emit('hide')")
      .q-dialog__inner.flex.no-pointer-events.q-dialog__inner--standard.fixed-full.flex-center(tabindex="-1")
        .q-pa-md.bg-grey-9
          slot
</template>

<script>
export default {
  name: 'Dialog',
  components: {},
  props: {
    value: {
      type: Boolean,
      required: true
    },
  },
  data () {
    return {
    }
  },
  computed: {},
  watch: {},
  created () {},
  mounted () {},
  updated () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .slide-enter-active, .slide-leave-active
    transition: transform .3s ease-in-out

  .slide-enter
    transform: translate3d(0, 100%, 0)
  .slide-leave-to
    transform: translate3d(0, 100%, 0)
</style>
