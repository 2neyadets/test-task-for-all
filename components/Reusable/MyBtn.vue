<template lang="pug">
  button.my-btn(
    @click="$emit('click')"
    :ref="'button' + id"
    :type="type"
    role="button"
    tabindex="0"
    :class="btnClass"
    :disabled="disable"
  )
    .full-height.row.items-center(:style="`font-size: ${size}px;`")
      template(v-if="icon")
        .col-auto.flex.flex-center
          i.material-icons(
            :class="label ? 'margin-right' : ''"
          ) {{icon}}
      .col {{label}}
</template>

<script>
import { getId } from '../../assets/helpers'

export default {
  name: 'MyBtn',
  components: {},
  props: {
    type: {
      type: String,
      default: 'button',
    },
    label: {
      type: String,
      default: '',
    },
    icon: {
      type: String,
      default: '',
    },
    size: {
      type: [String, Number],
      default: 20,
    },
    round: {
      type: Boolean,
      default: false,
    },
    active: {
      type: Boolean,
      default: false,
    },
    flat: {
      type: Boolean,
      default: false,
    },
    disable: {
      type: Boolean,
      default: false,
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
  },
  data () {
    return {
      id: getId(),
    }
  },
  computed: {
    btnClass () {
      return {
        'my-btn--hoverable': !this.disable,
        'my-btn--flat': this.flat,
        'my-btn--round': this.round,
        'my-btn--active': this.active,
      }
    },
  },
  watch: {},
  created () {},
  mounted () {
    if (this.autofocus) this.$refs['button' + this.id].focus()
  },
  updated () {},
  methods: {},
}
</script>

<style scoped lang="stylus">
  .my-btn
    display inline-flex
    flex-direction column
    align-items stretch
    position relative
    outline 0
    border 1px solid white
    vertical-align middle
    padding 6px
    font-size 14px
    line-height 1.715em
    text-decoration none
    color white
    background transparent
    font-weight 500
    text-transform uppercase
    text-align center
    width auto
    height auto
    transition .5s

    &--hoverable:hover, &--hoverable:focus
      color black
      background white
    &--flat
      border 0
      &:hover, &:focus
        color inherit
        background rgba(255, 255, 255, .15)
    &--round
      border-radius 50%
    &--active
      color black
      background white
  .margin-right
    margin-right 12px
</style>
