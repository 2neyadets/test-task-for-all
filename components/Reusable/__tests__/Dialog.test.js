import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Dialog from '../Dialog'
import MyBtn from '../MyBtn'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Dialog component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Dialog, {
      localVue,
      propsData: {
        value: true,
      },
    })
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick, если нужно что-то выпонять внутри него
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('disappear when false value', async () => {
    expect.assertions(2)
    wrapper.setProps({
      value: false,
    })
    await localVue.nextTick()
    expect(wrapper.find('.dialog__backdrop').exists()).toBe(false)
    expect(wrapper.find('.dialog__inner').exists()).toBe(false)
  })

  it('can emit "hide" on backdrop click', () => {
    expect.assertions(1)
    wrapper.find('.dialog__backdrop').trigger('click')
    expect(wrapper.emitted().hide[0]).toEqual([])
  })

  it('slot works', () => {
    expect.assertions(1)
    const localWrapper = shallowMount(Dialog, {
      propsData: {
        value: true,
      },
      slots: {
        default: MyBtn,
      }
    })
    expect(localWrapper.find('button').exists()).toBe(true)
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
