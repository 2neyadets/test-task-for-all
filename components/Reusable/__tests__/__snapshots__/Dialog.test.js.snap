// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog component match snapshot 1`] = `
<transition-stub
  name="slide"
>
  <div
    class="dialog fullscreen no-pointer-events"
  >
    <div
      class="dialog__backdrop fixed-full"
    />
    <div
      class="dialog__inner flex no-pointer-events fixed-full flex-center"
      tabindex="-1"
    >
      <div
        class="pa-md bg-grey-9"
      />
    </div>
  </div>
</transition-stub>
`;
